### ISSUE TITLE
The SSL handshake failure made a re-appearance

### ISSUE THREAD

**Issue by yaronyg** [thaliproject/Thali_CordovaPlugin#1406]
**Repro**: 
1. Clone aaladev_1356_yarong
2. Set 'jx' to point at v0.3.1.5 Spider Monkey (NOT V8! V8 crashes instantly)
3. After doing usual set up go to test/www/jxcore and run 'jx runCoordinatedTests.js --filter bv_tests_/testThaliManagerCoordinated.js'

If you do this a bunch of times then eventually you'll see:

```
2016-10-21 22:51:28 - DEBUG thaliReplicationPeerAction: 'Got error on replication -  500 140735249612800:error:14077102:SSL routines:SSL23_GET_SERVER_HELLO:unsupported protocol:../deps/openssl/openssl/ssl/s23_clnt.c:726:
'

2016-10-21 22:51:28 - DEBUG thaliPeerPoolOneAtATime: 'action returned with error from startError: 140735249612800:error:14077102:SSL routines:SSL23_GET_SERVER_HELLO:unsupported protocol:../deps/openssl/openssl/ssl/s23_clnt.c:726:
 - Action ID: 25, Action Type: ReplicationAction, Connection Type: tcp, Peer Identifier: BLEA5pScH_iqY_srYBZkpa9C_Qa-FfjYjDXa6JikgvaLw6IdCjOqhaTIsatheG6lhSbiHnxCof2VNKYv0KbaXIs'
```

**Comment by yaronyg** [comment]
We are getting closer to figuring this out. We can see that this is the result of a data pass from JXcore to OpenSSL. Aren't clear if this is a memory corruption or a code flaw.

**Comment by enricogior** [comment]
The SecureContext is created in jxcore:

https://github.com/thaliproject/jxcore/blob/master/src/wrappers/node_crypto.cc#L159-L165
```
JS_METHOD(SecureContext, New) {
  SecureContext* p = new SecureContext();
  JS_CLASS_NEW_INSTANCE(obj, SecureContext);
  p->Wrap(obj);
  RETURN_POINTER(obj);
}
JS_METHOD_END
```

The SecureContext is associated to the connection:

https://github.com/thaliproject/jxcore/blob/master/src/wrappers/node_crypto.cc#L1223-L1242
```
JS_METHOD(Connection, New) {
  Connection* p = new Connection();
  JS_CLASS_NEW_INSTANCE(obj, Connection);

  p->Wrap(obj);

  if (args.Length() < 1 || !args.IsObject(0)) {
    THROW_EXCEPTION("First argument must be a crypto module Credentials");
  }

  SecureContext* sc =
      ObjectWrap::Unwrap<SecureContext>(JS_VALUE_TO_OBJECT(GET_ARG(0)));

  bool is_server = args.GetBoolean(1);

  p->ssl_ = SSL_new(sc->ctx_);
  p->bio_read_ = BIO_new(BIO_s_mem());
  p->bio_write_ = BIO_new(BIO_s_mem());

  SSL_set_app_data(p->ssl_, p);

  ...

}
JS_METHOD_END
```

The SecureContext is passed to SSL_connect:

https://github.com/thaliproject/jxcore/blob/master/src/wrappers/node_crypto.cc#L1716-L1735
```
JS_METHOD(Connection, Start) {
  Connection* ss = Connection::Unwrap(args.GetHolder());

  if (!SSL_is_init_finished(ss->ssl_)) {
    int rv;
    if (ss->is_server_) {
      rv = SSL_accept(ss->ssl_);
      ss->HandleSSLError("SSL_accept:Start", rv, kZeroIsAnError, kSyscallError);
    } else {
      rv = SSL_connect(ss->ssl_);
      ss->HandleSSLError("SSL_connect:Start", rv, kZeroIsAnError,
                         kSyscallError);
    }

    RETURN_PARAM(STD_TO_INTEGER(rv));
  }

  RETURN_PARAM(STD_TO_INTEGER(0));
}
JS_METHOD_END
```

The SecureContext passed to SSL_connect was retrieved using args.GetHolder() to 
get the connection object:
 
https://github.com/thaliproject/jxcore/blob/master/src/jx/Proxy/Mozilla_340/PArguments.cc#L22-L26
```
void *PArguments::GetHolder() {
  MozJS::Value val(JS_THIS(ctx_, jsval_), ctx_);

  return val.GetPointerFromInternalField(0);
}
```

GetHolder calls GetPointerFromInternalField that calls the SM API JS_GetPrivate:

https://github.com/thaliproject/jxcore/blob/master/src/jx/Proxy/Mozilla_340/MozJS/MozValue.cc#L590-L609
```
void *Value::GetPointerFromInternalField(const int index) {
  INNER_VALUE_TO_OBJECT(object_, false);
  JSObject *object_ = object_rt;

  JS::RootedObject ret_val(ctx_);
  if (object_ != nullptr) {
    if (!JS_HasReservedSlot(object_, GC_SLOT_GC_CALL) &&
        !JS_HasPrivate(object_)) {
      // perhaps this object was created on JS land
      // add a property that can trigger the event when the base
      // object is GC'ed
      Natify(object_rt, false, &ret_val);
      object_ = ret_val;
    }
  }

  if (object_ == nullptr) return nullptr;

  return JS_GetPrivate(object_);
}
```

https://github.com/thaliproject/jxcore/blob/master/deps/mozjs/src/jsapi.cpp#L2269-L2274
```
JS_PUBLIC_API(void *)
JS_GetPrivate(JSObject *obj)
{
    /* This function can be called by a finalizer. */
    return obj->getPrivate();
}
```
The SecureContext retrieved from the unwrapped object (returned by SM) contains the corrupted data.

**Comment by enricogior** [comment]
When the SSL error occurs, the following connection attempts succeed, the reason is that after the error the connection is closed and new connections are created, with a new connection the Secure Context is recreated.

**Comment by yaronyg** [comment]
The PSK code was always passing values to the TLS layer wrong. But by a miracle most of the time this didn't cause a failed connection, but sometimes it does. So we have to fix the PSK code.

**Comment by enricogior** [comment]
https://github.com/thaliproject/Thali_CordovaPlugin/blob/aaladev_1356_yarong/thali/NextGeneration/replication/thaliReplicationPeerAction.js#L207-L220

Thali uses the `secureOptions` attribute assuming that jxcore won't use it, but it does.

**Comment by yaronyg** [comment]
So the bug is caused by our work around to thaliproject/Thali_CordovaPlugin#730 so we have to fix that bug.

**Comment by enricogior** [comment]
@yaronyg when #730 is fixed we need to test again jxcore with the same test, since during debugging we noticed two things that we don't know if they were caused by #730:
 - in the SSL handshaking there were too many requests coming from the client (I don't think it's caused by #730 but we didn't really investigate it so it's just a guess that is not related to #730).
 - the secureOptions attribute was always overwritten with random values, most of the times the values were not causing the SSL error just by chance, but in one particular case the value was way off and causing the SSL negotiation to fail. We need to make sure that the value that was causing the error was indeed produces by #730 and not by another bug that somehow is hidden by the #730 issue.
